core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -62,34 +66,108 @@ public void init() {
    }
 
    @Override
-   public final void wrapEntryForReading(InvocationContext ctx, Object key, int segment, boolean isOwner) {
+   public final CompletionStage<Void> wrapEntryForReading(InvocationContext ctx, Object key, int segment, boolean isOwner) {
       if (!isOwner && !isL1Enabled) {
-         return;
+         return CompletableFutures.completedNull();
       }
+      CompletionStage<Void> returnedStage = CompletableFutures.completedNull();
       CacheEntry cacheEntry = getFromContext(ctx, key);
       if (cacheEntry == null) {
-         cacheEntry = getFromContainerForRead(key, segment, isOwner);
+         InternalCacheEntry readEntry = getFromContainer(key, segment);
+         CompletionStage<Boolean> expiredStage = null;
+         if (readEntry == null) {
+            if (isOwner) {
+               addReadEntryToContext(ctx, NullCacheEntry.getInstance(), key);
+            }
+         } else if (isOwner || readEntry.isL1Entry()) {
+            cacheEntry = readEntry;
```

Comment:
I think you can use `readEntry` directly; and you can make if `final` and remove `finalEntry`

Reply:
Yes, originally I was trying to do that, but had some issues and forgot to clean it up. It is much easier than this now :)